Properly Gate the Paid Newsletter Feature based on Newsletter Config

romarioraffington opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes https://github.com/Automattic/jetpack/issues/29693
Fixes https://github.com/Automattic/jetpack/issues/29657

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
Previously we were gating the frontend components and backend code based on the following criteria:

1. Newsletter feature is enabled
2. Newsletter Plan is configured
3. Stripe KYC completed.

We need the above checks for some frontend components while other components should be displayed when the Newsletter feature is enabled. This PR resolves the issue by separating the configuration check from the feature enable check and using each check to gate components where required.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

### Before testing

1. Create a new test site on WPCOM and sandbox that test site
2. Ensure that the test site is on a  personal plan or highee
3. Ensure that you are sandboxing store by adding the following to `0-sandbox.php`:

```
define( 'USE_STORE_SANDBOX', true );
define( 'USE_BILLING_SANDBOX', true );
define( 'BILLINGDADDY_SANDBOX', rtrim( 'https://' . `hostname` ) );
```

4. Ensure that you are sandboxing  `public-api.wordpress.com`
5. Add this PR to your sandbox: `jetpack rsync jetpack [sandbox_url]:~/public_html/wp-content/mu-plugins/jetpack-plugin/production`

### Test Case A

1. Add the Subscribe block to a post
2. Verify the "Add Payments" button is displayed on the Subscribe block and in the post publish panel

### Test Case B

1. Publish a post with Paid Subscriber's access level. This can be done from the post publish panel
2. Verify the access level is saved correctly
3. Open the post in an incognito tab and verify that the post is locked

romarioraffington committed:
Create a different check for newsletter config

romarioraffington committed:
Gate components based on Newsletter config

romarioraffington committed:
Revert to running newsletter code when feature enabled

romarioraffington committed:
changelog

n3f commented:
Using "real" paid sites (i used credits for an alternate account).
* `A` didn't work for me on a personal plan. (neffff59d89e8b58feaab74fff.wordpress.com)
	* it did show the number of subscribers though (weird - should only be myself - maybe that's expected).
* `B` doesn't even show the paid subscribers panel on the post sidebar. 😕
* The calypso interface is coming through wordpress.com -- not sure if something else needs to be sandboxed.
* Saving pages didn't appear to work either... tried changing browsers... non beuno
* Put it into sandboxed mode... 
  * `A` worked
  * publishing didn't... after reloading the posts page it looked like it saved, but attempting to go to it I got a FATAL: `Fatal error: Uncaught Error: Call to undefined method Jetpack_Gutenberg::is_newsletter_enabled() in /home/wpcom/public_html/wp-content/mu-plugins/jetpack-plugin/production/extensions/blocks/subscriptions/subscriptions.php:371`
  * `B` still not working (still not in the sidebar...)
* Trying premium site:
  * `A` worked. Then I tried connecting stripe -- i missed that above, maybe that was the issue.
  * After connecting stripe it didn't show the plan, but the url *did* have the `?stripe_connect_success=earn-newsletter` url param...
  * pages not publishing...
  * `B` still not working... on pages 🤦 , can see it on posts.
  * Viewing the subscriber only post in incognito shows a fatal like above.
  * Can't edit "saved" pages it's just a blank page. I'm not actually sure that they saved.  Looks like something didn't go right.
* trying business plan:
  * it's different everytime.  After I connect stripe I got a weird notice after adding the subscribe block that a payment plan was added -- but I haven't done that yet. and `A` isn't working.
  * The `newsletter` menu is missing in the post sidebar.

artpi commented on projects/plugins/jetpack/class.jetpack-gutenberg.php:
```diff
@@ -715,7 +715,16 @@ public static function enqueue_block_editor_assets() {
 				 * @param bool false Enable the Paid Newsletters feature in the block editor context.
 				 */
-				'is_newsletter_feature_enabled' => self::is_newsletter_enabled(),
+				'is_newsletter_feature_enabled' => apply_filters( 'jetpack_subscriptions_newsletter_feature_enabled', false ),
```
```php
				'is_newsletter_feature_enabled' => ( apply_filters( 'jetpack_subscriptions_newsletter_feature_enabled', false ) && class_exists( '\Jetpack_Memberships' )  );

```

artpi commented on projects/plugins/jetpack/class.jetpack-gutenberg.php:
```diff
@@ -715,7 +715,16 @@ public static function enqueue_block_editor_assets() {
 				 * @param bool false Enable the Paid Newsletters feature in the block editor context.
 				 */
-				'is_newsletter_feature_enabled' => self::is_newsletter_enabled(),
+				'is_newsletter_feature_enabled' => apply_filters( 'jetpack_subscriptions_newsletter_feature_enabled', false ),
```
```suggestion
				'is_newsletter_feature_enabled' => ( apply_filters( 'jetpack_subscriptions_newsletter_feature_enabled', false ) && class_exists( '\Jetpack_Memberships' ) ),
```

romarioraffington replied:
Fixed in a915233d7b4a78ad972aad2b6f3c27f971d37160

artpi commented on projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php:
```diff
@@ -58,8 +58,14 @@ function register_block() {
+		/** This filter is documented in class.jetpack-gutenberg.php */
+		! apply_filters( 'jetpack_subscriptions_newsletter_feature_enabled', false )
+	) {
+		return;
```
Interesting. So it wasn't saving because it was not registering a meta here?

romarioraffington replied:
Still trying to figure out what's happening here. I reverted to the previous code as this code wasn't being run when we needed it to

artpi reviewed (changes requested):
You still have a use of the is_newsletter_enabled in https://github.com/Automattic/jetpack/pull/29710/files#diff-9164d71f603706b3c6cc58c71ba3f2fbb12b1dba772adca76793f60c24ddbca7R371

When you publish a post, this is on the frontend:
```

Fatal error: Uncaught Error: Call to undefined method Jetpack_Gutenberg::is_newsletter_enabled() in /home/wpcom/public_html/wp-content/mu-plugins/jetpack-plugin/production/extensions/blocks/subscriptions/subscriptions.php:371 Stack trace: #0 /home/wpcom/public_html/wp-includes/class-wp-block.php(263): Automattic\Jetpack\Extensions\Subscriptions\render_block(Array, '', Object(WP_Block)) #1 /home/wpcom/public_html/wp-includes/class-wp-block.php(246): WP_Block->render() #2 /home/wpcom/public_html/wp-includes/blocks.php(1051): WP_Block->render() #3 /home/wpcom/public_html/wp-includes/blocks.php(1089): render_block(Array) #4 /home/wpcom/public_html/wp-content/plugins/gutenberg-core/v15.4.0/build/block-library/blocks/pattern.php(41): do_blocks('\n<!-- wp:group ...') #5 /home/wpcom/public_html/wp-includes/class-wp-block.php(263): gutenberg_render_block_core_pattern(Array, '', Object(WP_Block)) #6 /home/wpcom/public_html/wp-includes/class-wp-block.php(246): WP_Block->render() #7 /home/wpcom/public_html/wp-includes/blocks.php(1051): WP_Block->render() #8 /home/wpcom/public_html/wp-includes/blocks.php(1089): render_block(Array) #9 /home/wpcom/public_html/wp-content/plugins/gutenberg-core/v15.4.0/build/block-library/blocks/pattern.php(41): do_blocks('\n<!-- wp:group ...') #10 /home/wpcom/public_html/wp-includes/class-wp-block.php(263): gutenberg_render_block_core_pattern(Array, '', Object(WP_Block)) #11 /home/wpcom/public_html/wp-includes/blocks.php(1051): WP_Block->render() #12 /home/wpcom/public_html/wp-includes/blocks.php(1089): render_block(Array) #13 /home/wpcom/public_html/wp-content/plugins/gutenberg-core/v15.4.0/build/block-library/blocks/template-part.php(147): do_blocks('<!-- wp:pattern...') #14 /home/wpcom/public_html/wp-content/mu-plugins/security.php(2918): gutenberg_render_block_core_template_part(Array) #15 /home/wpcom/public_html/wp-includes/class-wp-block.php(263): wpcom_safe_gutenberg_render_block_core_template_part(Array, '', Object(WP_Block)) #16 /home/wpcom/public_html/wp-includes/blocks.php(1051): WP_Block->render() #17 /home/wpcom/public_html/wp-includes/blocks.php(1089): render_block(Array) #18 /home/wpcom/public_html/wp-includes/block-template.php(240): do_blocks('<!-- wp:templat...') #19 /home/wpcom/public_html/wp-includes/template-canvas.php(12): get_the_block_template_html() #20 /home/wpcom/public_html/wp-includes/template-loader.php(106): include('/home/wpcom/pub...') #21 /home/wpcom/public_html/wp-blog-header.php(19): require_once('/home/wpcom/pub...') #22 /home/wpcom/public_html/index.php(17): require('/home/wpcom/pub...') #23 {main} thrown in /home/wpcom/public_html/wp-content/mu-plugins/jetpack-plugin/production/extensions/blocks/subscriptions/subscriptions.php on line 371

Fatal error: Uncaught Error: Call to undefined method Jetpack_Gutenberg::is_newsletter_enabled() in /home/wpcom/public_html/wp-content/mu-plugins/jetpack-plugin/production/extensions/blocks/subscriptions/subscriptions.php:371 Stack trace: #0 /home/wpcom/public_html/wp-includes/class-wp-block.php(263): Automattic\Jetpack\Extensions\Subscriptions\render_block(Array, '', Object(WP_Block)) #1 /home/wpcom/public_html/wp-includes/class-wp-block.php(246): WP_Block->render() #2 /home/wpcom/public_html/wp-includes/blocks.php(1051): WP_Block->render() #3 /home/wpcom/public_html/wp-includes/blocks.php(1089): render_block(Array) #4 /home/wpcom/public_html/wp-content/plugins/gutenberg-core/v15.4.0/build/block-library/blocks/pattern.php(41): do_blocks('\n
```

romarioraffington committed:
Add Jetpack_Memberships to is_newsletter_feature_enabled check

romarioraffington committed:
Fix fatal error

romarioraffington commented:
> You still have a use of the is_newsletter_enabled in https://github.com/Automattic/jetpack/pull/29710/files#diff-9164d71f603706b3c6cc58c71ba3f2fbb12b1dba772adca76793f60c24ddbca7R371

Thanks for catching this. Fixed in [d024604](https://github.com/Automattic/jetpack/pull/29710/commits/d0246048d17884d7c90ca96172008c453bf80911)

n3f reviewed (dismissed):
Looks good to me now and it's working...

romarioraffington committed:
Merge branch 'trunk' into fix/display-add-payments-button-for-subscribe-block

jeherve committed:
Fix comments

We used $$next_version$$ instead of $$next-version$$, which is not supported by the script.

jeherve reviewed (approved):
This should work. 🚢
